File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    
    string longestPalindrome(string s) {
        int best_left = 0, best_right = 0;
        vector<vector<int>> dp(s.size(), vector<int>(s.size(), 0));
        
        for (int index = 0; index < s.size(); index ++){
            dp[index][index] = 1;
        }
        
        for (int row = 1; row < s.size(); row ++ ){
            for (int col = 0; col < row; col ++){
                if (row - 1 == col){
                    dp[row][col] = (s[row] == s[col]);
                }
                else{
                    dp[row][col] = dp[row - 1][col + 1] && s[row] == s[col];
                }
                
                if (dp[row][col] == 1){
                    if(best_right - best_left < row - col){
                        best_right = row;
                        best_left = col;
                    }
                }
            }
        }
        return s.substr(best_left, best_right - best_left + 1);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[0]*(len(s)+1) for _ in range(len(s) + 1)]  
        best_left = best_right = cur_answer = 0
        for index in range(len(s)):
            dp[index][index] = 1
        for step in range(1, len(s)):
            for left in range(len(s)-step):
                shift = step + left
                if s[shift] == s[left] and dp[left+1][shift-1] == shift - left - 1:
                    dp[left][shift] = 2 + dp[left + 1][shift - 1]
                if dp[left][shift] > cur_answer:
                    cur_answer = dp[left][shift]
                    best_left = left
                    best_right = shift
                    
        return s[best_left:best_right+1]
        